observability/src/main/java/io/grpc/observability/interceptors/GcpLogHelper.java

```diff
@@ -0,0 +1,430 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.observability.interceptors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Charsets;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Duration;
+import com.google.protobuf.util.Timestamps;
+import io.grpc.Attributes;
+import io.grpc.Deadline;
+import io.grpc.Grpc;
+import io.grpc.InternalMetadata;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor.Marshaller;
+import io.grpc.Status;
+import io.grpc.internal.TimeProvider;
+import io.grpc.observability.logging.Sink;
+import io.grpc.observabilitylog.v1.GrpcLogRecord;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.Address;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventType;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.LogLevel;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Helper class for GCP observability logging.
+ */
+final class GcpLogHelper {
+  private static final Logger logger = Logger.getLogger(GcpLogHelper.class.getName());
+  static final Metadata.Key<byte[]> STATUS_DETAILS_KEY =
+      Metadata.Key.of(
+          "grpc-status-details-bin",
+          Metadata.BINARY_BYTE_MARSHALLER);
+
+  /**
+   *  Class for proto related helper methods.
+   */
+  static final class GcpLogSinkWriter {
+    private final Sink sink;
+    private final TimeProvider timeProvider;
+
+    GcpLogSinkWriter(Sink sink, TimeProvider timeProvider) {
+      this.sink = sink;
+      this.timeProvider = timeProvider;
+    }
+
+    GrpcLogRecord.Builder createTimestamp() {
+      long nanos = timeProvider.currentTimeNanos();
+      return GrpcLogRecord.newBuilder().setTimestamp(Timestamps.fromNanos(nanos));
+    }
+
+
+    /**
+     * Logs the client header.
+     */
+    void logClientHeader(
+        long seqId,
+        String serviceName,
+        String methodName,
+        @Nullable String authority,
+        @Nullable Duration timeout,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_SERVER,
+          "peerAddress can only be specified by server");
```

Comment:
Can we add a comment explaining why it is so i.e. "peerAddress can only be specified by server"? I think that's because from the client the peerAddress is not yet known (because the RPC is not yet assigned to a transport). Whereas on the server the RPC has arrived from a peer so peerAddress is known?

Reply:
As per our discussion offline, leaving it as. Based on what we decide, I will make change when implementing server interceptor.